sympy/combinatorics/fp_groups.py

```diff
@@ -0,0 +1,336 @@
+# H: subgroup of fp_grp.
+# NOTE: We start with H as being only a list of words in generators
+#       of "fp_grp". Since `.subgroup` method has not been implemented.
+
+def coset_table(fp_grp, H):
+    # list of generators and their inverses, should be made a tuple.
+    A = []
+    for gen in fp_grp:
+        A.extend([gen, gen.inverse()])
+    r = len(A)
+    # start with table of 1 row (can change) and 2*(no. of gens) coloumns (fixed)
+    table = [r*[-1]]
+
+# coset_table: Mathematically a coset table
+#               represented using a list of lists
+# alpha: Mathematically a coset (precisely, a live coset)
+#       represented by an integer between i with 1 <= i <= n
+#       α ∈ c
+# x: Mathematically an element of "A" (set of generators and
+#   their inverses), represented using "FpGroupElm"
+
+def define(C, alpha, x):
+    # now C.n is now obtainable
+#    if C.n == DefaultMaxLimit:
+#        # abort the further generation of cosets
+#        return
+#    C.n += 1
+    # beta is the new coset generated
+    beta = n
+    C.p[beta] = beta
+    C[alpha][C.A.index(x)] = beta
+    C[alpha][C.A.index(x.inverse())] = alpha
+
+
+def scan(C, alpha, word, A):
+    """
+    >>> F, x, y = free_group("x, y")
+    >>> A = [x, x**-1, y, y**-1]
+
+    # Example 5.1, Pg 150
+    >>> c = [[0, 0, 1, 2], [-1, -1, -1, 0], [-1, -1, 0, -1]]
+    >>> scan(c, 1, y**3, A)
+    # this lead to deductions 2^y = 3, 3^(y^-1) = 2 (1, 2, 3 represent cosets in this comment)
+    >>> c
+    [[0, 0, 1, 2], [-1, -1, 2, 0], [-1, -1, 0, 1]]
+    >>> scan(c, 0, x**-1*y**-1*x*y, A)
+    >>> c
+    [[0, 0, 1, 2], [-1, -1, 2, 0], [2, 2, 0, 1]]
+    >>> scan(c, 1, x**-1*y**-1*x*y, A)
+    >>> c
+    [[0, 0, 1, 2], [1, 1, 2, 0], [2, 2, 0, 1]]
+
+    # Example 5.2, Pg 154
+    >>> c1 = [[1, 1, -1, -1], [0, 0, -1, -1]]
+    >>> scan(c1, 0, x*y, A)
+    >>> c1
+    [[1, 1, -1, 1], [0, 0, 0, -1]]
+
+    >>> c2 = [[1, 1, 2, 1], [0, 0, 0, -1], [-1, -1, -1, 0]]
+    >>> scan(c2, 1, y**3, A)
+    >>> c2
+    [[1, 1, 2, 1], [0, 0, 0, 2], [-1, -1, 1, 0]]
+
+    >>> c3 = [[1, 1, 2, 1], [0, 0, 0, 2], [3, 3, 1, 0], [2, 2, -1, -1]]
+    >>> scan(c4, 2, (x*y)**3, A)
+    >>> c3
+    [[1, 1, 2, 1], [0, 0, 0, 2], [3, 3, 1, 0], [2, 2, 3, 3]]
+
+    """
+    # alpha is an integer representing a "coset"
+    # since scanning can be in two cases
+    # 1. for alpha=0 and w in Y (i.e generating set of H)
+    # 2. alpha in sigma (set of live cosets), w in R (relators)
+    f = alpha
+    i = 0
+    r = len(word)
+    # list of union of generators and their inverses
+    # A = coset_table.A
+    while i < r and C[f][A.index(word.subword(i, i+1))] != -1:
+        f = C[f][A.index(word.subword(i, i+1))]
+        i += 1
+    if i >= r:
+        if f != alpha:
+            # implement the "coincidence" routine on Pg 158 of Handbook.
+            coincidence(C, f, alpha)
+            return
+    b = alpha
+    j = r
+    while j >= i and C[b][A.index(word.subword(j-1, j).inverse())] != -1:
```

Comment:
I am looking at the Pg. 155 of Derek Holt, pseudo-code for scan line number 9-10 there is "b​^{x^​−​​1}"​​ is defined"
According to the Derek Holt. pseudo-code is
```
while j >= i and b^{x^-1} is defined
    do; b:=b^{x^-1}; j:=j–1;
```

My point is in book it uses `b^{x^-1}`, which I initially fortunately/unfortunately coded as it is my commit right now. Also what we have commited here seems to work, but the one in book doesn't do the work.
You can test them using the "docs", first example.